packages/salesforcedx-vscode-core/src/commands/isvdebugging/bootstrapCmd.ts

```diff
@@ -64,52 +78,278 @@ export class IsvDebugBootstrapExecutor extends SfdxCommandletExecutor<{}> {
+      )
+      .withArg('force:package:installed:list')
+      .withFlag('-u', data.sessionId)
+      .withArg('--json')
+      .build();
+  }
+
+  public buildRetrievePackagesSourceCommand(
+    data: IsvDebugBootstrapConfig,
+    packageNames: string[]
+  ): Command {
+    return new SfdxCommandBuilder()
+      .withDescription(
+        nls.localize('isv_debug_bootstrap_step6_retrieve_packages_source')
+      )
+      .withArg('force:mdapi:retrieve')
+      .withFlag('-r', this.relativeMetdataTempPath)
+      .withFlag('-p', packageNames.join(','))
+      .withFlag('-u', data.sessionId)
+      .build();
+  }
+
+  public buildMetadataApiConvertPackageSourceCommand(
+    packageName: string
+  ): Command {
+    return new SfdxCommandBuilder()
+      .withDescription(
+        nls.localize(
+          'isv_debug_bootstrap_step7_convert_package_source',
+          packageName
+        )
+      )
+      .withArg('force:mdapi:convert')
+      .withFlag(
+        '-r',
+        path.join(this.relativeMetdataTempPath, 'packages', packageName)
+      )
+      .withFlag('-d', path.join('packages', packageName))
+      .build();
+  }
+
+  public parsePackageInstalledListJson(packagesJson: string): string[] {
+    const packagesData = JSON.parse(packagesJson);
+    return packagesData.result.map((entry: any) => entry.SubscriberPackageName);
+  }
+
+  public async execute(
+    response: ContinueResponse<IsvDebugBootstrapConfig>
+  ): Promise<void> {
     const cancellationTokenSource = new vscode.CancellationTokenSource();
     const cancellationToken = cancellationTokenSource.token;
 
     const projectParentPath = response.data.projectUri;
     const projectPath = path.join(projectParentPath, response.data.projectName);
+    const projectMetadataTempPath = path.join(
+      projectPath,
+      this.relativeMetdataTempPath
+    );
+    const apexRetrievePackageXmlPath = path.join(
+      projectPath,
+      this.relativeApexPackageXmlPath
+    );
 
-    const createProjectExecution = new CliCommandExecutor(
+    // 1: create project
+    await this.executeCommand(
       this.buildCreateProjectCommand(response.data),
-      {
-        cwd: projectParentPath
-      }
-    ).execute(cancellationToken);
-
-    createProjectExecution.processExitSubject.subscribe(async data => {
-      if (data != undefined && data.toString() === '0') {
-        const configureProjectExecution = new CliCommandExecutor(
-          this.buildConfigureProjectCommand(response.data),
-          {
-            cwd: projectPath
-          }
-        ).execute(cancellationToken);
-
-        configureProjectExecution.processExitSubject.subscribe(async data2 => {
-          if (data2 != undefined && data2.toString() === '0') {
-            // last step is open the folder
-            await vscode.commands.executeCommand(
-              'vscode.openFolder',
-              vscode.Uri.parse(projectPath)
-            );
-          }
-        });
+      { cwd: projectParentPath },
```

Comment:
All the steps use the same options, cancellationTokenSource, and cancellationToken. Can `executeCommand()` use those as defaults?

Reply:
We could save them as members of the executor. I was waiting for the Uber command work to settle before refactoring further in this area.

Reply:
Über command work is done already per @Sweetman.